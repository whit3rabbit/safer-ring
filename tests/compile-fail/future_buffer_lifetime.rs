// This test should fail to compile because futures must not outlive their buffers

use safer_ring::{Ring, PinnedBuffer, ReadFuture};

fn create_future() -> ReadFuture<'static, 'static> {
    let ring = Ring::new(10).unwrap();
    let mut buffer = PinnedBuffer::with_capacity(1024);
    
    // This should not compile - future cannot have 'static lifetime
    // when buffer and ring have local lifetimes
    ring.read(0, buffer.as_mut_slice()).unwrap()
}

fn main() {
    let _future = create_future();
}