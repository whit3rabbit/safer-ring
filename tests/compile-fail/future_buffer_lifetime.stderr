error[E0596]: cannot borrow `ring` as mutable, as it is not declared as mutable
  --> tests/compile-fail/future_buffer_lifetime.rs:11:5
   |
11 |     ring.read(0, buffer.as_mut_slice()).unwrap()
   |     ^^^^ cannot borrow as mutable
   |
help: consider changing this to be mutable
   |
6  |     let mut ring = Ring::new(10).unwrap();
   |         +++

error[E0515]: cannot return value referencing local variable `buffer`
  --> tests/compile-fail/future_buffer_lifetime.rs:11:5
   |
11 |     ring.read(0, buffer.as_mut_slice()).unwrap()
   |     ^^^^^^^^^^^^^------^^^^^^^^^^^^^^^^^^^^^^^^^
   |     |            |
   |     |            `buffer` is borrowed here
   |     returns a value referencing data owned by the current function
